gravitee-apim-repository/gravitee-apim-repository-mongodb/src/main/java/io/gravitee/repository/mongodb/management/internal/key/ApiKeyMongoRepositoryImpl.java

```diff
@@ -37,38 +39,86 @@
+            );
+        } else if (filter.getExpireAfter() > 0) {
+            pipeline.add(match(gte("expireAt", new Date(filter.getExpireAfter()))));
+        } else if (filter.getExpireBefore() > 0) {
+            pipeline.add(match(lte("expireAt", new Date(filter.getExpireBefore()))));
+        }
 
-            query.addCriteria(expireAtCriteria);
+        if (filter.getPlans() != null) {
+            pipeline.add(lookup("subscriptions", "subscriptions", "_id", "sub"));
+            pipeline.add(unwind("$sub"));
+            pipeline.add(match(in("sub.plan", filter.getPlans())));
         }
 
-        query.with(Sort.by(Sort.Direction.DESC, "updatedAt"));
+        pipeline.add(sort(Sorts.descending("updatedAt")));
+
+        AggregateIterable<Document> aggregate = mongoTemplate
+            .getCollection(mongoTemplate.getCollectionName(ApiKeyMongo.class))
+            .aggregate(pipeline);
 
-        return mongoTemplate.find(query, ApiKeyMongo.class);
+        return getListFromAggregate(aggregate);
+    }
+
+    @Override
+    public List<ApiKeyMongo> findByKeyAndApi(String key, String api) {
+        List<Bson> pipeline = List.of(
+            match(eq("key", key)),
+            lookup("subscriptions", "subscriptions", "_id", "sub"),
+            unwind("$sub"),
+            match(eq("sub.api", api))
+        );
+
+        AggregateIterable<Document> aggregate = mongoTemplate
+            .getCollection(mongoTemplate.getCollectionName(ApiKeyMongo.class))
+            .aggregate(pipeline);
+
+        return getListFromAggregate(aggregate);
+    }
+
+    @Override
+    public List<ApiKeyMongo> findByPlan(String plan) {
+        List<Bson> pipeline = new ArrayList<>();
+        pipeline.add(lookup("subscriptions", "subscriptions", "_id", "sub"));
+        pipeline.add(unwind("$sub"));
+        pipeline.add(match(eq("sub.plan", plan)));
+
+        AggregateIterable<Document> aggregate = mongoTemplate
+            .getCollection(mongoTemplate.getCollectionName(ApiKeyMongo.class))
+            .aggregate(pipeline);
+
+        return getListFromAggregate(aggregate);
+    }
+
+    private List<ApiKeyMongo> getListFromAggregate(AggregateIterable<Document> aggregate) {
+        ArrayList<ApiKeyMongo> apiKeys = new ArrayList<>();
+        for (Document doc : aggregate) {
+            ApiKeyMongo apiKeyMongo = new ApiKeyMongo();
+            apiKeyMongo.setId(doc.getString("_id"));
+            apiKeyMongo.setKey(doc.getString("key"));
+            apiKeyMongo.setSubscriptions(new HashSet<>(doc.getList("subscriptions", String.class, List.of())));
+            apiKeyMongo.setApplication(doc.getString("application"));
+            apiKeyMongo.setCreatedAt(doc.getDate("createdAt"));
+            apiKeyMongo.setUpdatedAt(doc.getDate("updatedAt"));
+            apiKeyMongo.setRevoked(doc.getBoolean("revoked"));
+            apiKeyMongo.setPaused(doc.getBoolean("paused"));
+            apiKeys.add(apiKeyMongo);
```

Comment:
Instead of this inline mapping maybe you can use something like :
`ApiKeyMongo apiKeyMongo = mongoTemplate.getConverter().read(ApiKeyMongo.class, doc);`